cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -0,0 +1,466 @@
+    // todo use a wrapper that adds the new partition when the job is committed - that means we must serialize the
+    //      cconf into the hadoop conf to be able to instantiate this dataset in the output committer
+    return files.getOutputFormatClass();
+  }
+
+  @Override
+  public Map<String, String> getOutputFormatConfiguration() {
+    // all runtime arguments are passed on to the file set, so we can expect the partition time in the file set's
+    // output format configuration. If it is not there, the output format will fail to register this partition.
+    Map<String, String> config = files.getOutputFormatConfiguration();
+    Long time = TimePartitionedFileSetArguments.getOutputPartitionTime(runtimeArguments);
+    if (time == null) {
+      throw new DataSetException("Time must be given for the new output partition as a runtime argument.");
+    }
+    // add the output partition time to the output arguments of the embedded file set
+    Map<String, String> outputArgs = Maps.newHashMap();
+    outputArgs.putAll(config);
+    TimePartitionedFileSetArguments.setOutputPartitionTime(outputArgs, time);
+    return ImmutableMap.copyOf(outputArgs);
+  }
+
+  @Override
+  public FileSet getEmbeddedFileSet() {
+    return files;
+  }
+
+  @Override
+  public Map<String, String> getRuntimeArguments() {
+    return runtimeArguments;
+  }
+
+  //------ private helpers below here --------------------------------------------------------------
+
+  @VisibleForTesting
+  static byte[] generateRowKey(PartitionKey key, Partitioning partitioning) {
+    // validate partition key, convert values, and compute size of output
+    Map<String, FieldType> partitionFields = partitioning.getFields();
+    int totalSize = partitionFields.size() - 1; // one \0 between each of the fields
+    ArrayList<byte[]> values = Lists.newArrayListWithCapacity(partitionFields.size());
+    for (Map.Entry<String, FieldType> entry : partitionFields.entrySet()) {
+      String fieldName = entry.getKey();
+      FieldType fieldType = entry.getValue();
+      Comparable fieldValue = key.getField(fieldName);
+      if (fieldValue == null) {
+        throw new IllegalArgumentException(
+          String.format("Incomplete partition key: value for field '%s' is missing", fieldName));
+      }
+      if (!FieldTypes.validateType(fieldValue, fieldType)) {
+        throw new IllegalArgumentException(
+          String.format("Invalid partition key: value for %s field '%s' has incompatible type %s",
+                        fieldType.name(), fieldName, fieldValue.getClass().getName()));
+      }
+      byte[] bytes = FieldTypes.toBytes(fieldValue, fieldType);
+      totalSize += bytes.length;
+      values.add(bytes);
+    }
+    byte[] rowKey = new byte[totalSize];
+    int offset = 0;
+    for (byte[] bytes : values) {
+      System.arraycopy(bytes, 0, rowKey, offset, bytes.length);
+      offset += bytes.length;
+      if (offset < rowKey.length) {
```

Comment:
primitive array in Java is initialized with zeros. You can skip this logic and just have `offset += (bytes.length + 1)`.